pkg/bootstrap/platform/azure.go

```diff
@@ -0,0 +1,170 @@
+			return al.(string)
+		}
+		return ""
+	}
+	azureZoneFn = func(e *azureEnv) string {
+		if az, ok := e.computeMetadata["zone"]; ok {
+			return az.(string)
+		}
+		return ""
+	}
+)
+
+type azureEnv struct {
+	computeMetadata map[string]interface{}
+	networkMetadata map[string]interface{}
+}
+
+// IsAzure returns whether or not the platform for bootstrapping is Azure
+func IsAzure() bool {
+	sysVendor, err := ioutil.ReadFile(SysVendorPath)
+	if err != nil {
+		log.Warnf("Error reading sys_vendor in Azure platform detection: %v", err)
+	}
+	return strings.Contains(string(sysVendor), MicrosoftIdentifier)
+}
+
+// Attempts to update the API version
+func updateAPIVersion() {
+	updateVersionOnce.Do(func() {
+		bodyJson := stringToJson(getAPIVersions())
+		if newestVersions, ok := bodyJson["newest-versions"]; ok {
+			for _, version := range newestVersions.([]interface{}) {
+				if strings.Compare(version.(string), APIVersion) > 0 {
+					APIVersion = version.(string)
+				}
+			}
+		}
+	})
+}
+
+// NewAzure returns a platform environment for Azure
+func NewAzure() Environment {
+	updateAPIVersion()
+	e := &azureEnv{}
+	e.getMetadata()
+	return e
+}
+
+// Retrieves Azure instance metadata response body stores it in the Azure environment
+func (e *azureEnv) getMetadata() {
+	bodyJson := stringToJson(getAzureMetadata())
+	if computeMetadata, ok := bodyJson["compute"]; ok {
+		e.computeMetadata = computeMetadata.(map[string]interface{})
+	}
+	if networkMetadata, ok := bodyJson["network"]; ok {
+		e.networkMetadata = networkMetadata.(map[string]interface{})
+	}
+}
+
+// Generic Azure metadata GET request helper for the response body
+// Uses the default timeout for the HTTP get request
+func metadataRequest(query string) string {
+	client := http.Client{Timeout: defaultTimeout}
+	req, err := http.NewRequest("GET", fmt.Sprintf("%s?%s", InstanceUrl, query), nil)
+	if err != nil {
+		log.Warnf("Failed to create HTTP request: %v", err)
+		return ""
+	}
+	req.Header.Add("Metadata", "True")
+
+	response, err := client.Do(req)
+	if err != nil {
+		log.Warnf("HTTP request failed: %v", err)
+		return ""
+	}
+	if response.StatusCode != http.StatusOK {
+		log.Warnf("HTTP request unsuccessful with status: %v", response.Status)
+	}
+	defer response.Body.Close()
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		log.Warnf("Could not read response body: %v", err)
+		return ""
+	}
+	return string(body)
+}
+
+func stringToJson(s string) map[string]interface{} {
+	var sJson map[string]interface{}
+	if err := json.Unmarshal([]byte(s), &sJson); err != nil {
+		log.Warnf("Could not unmarshal response: %v:", err)
+	}
+	return sJson
+}
+
+// Returns Azure instance metadata. Must be run on an Azure VM
+func (e *azureEnv) Metadata() map[string]string {
+	md := map[string]string{}
+	for k, v := range azureTagsFn(e) {
+		md[k] = v
+	}
+	return md
+}
+
+// Locality returns the region and zone
+func (e *azureEnv) Locality() *core.Locality {
```

Comment:
Do we have similar helpers for other common metadata that will be helpful for routing? like labels, instance name, etc. i.e. can we transform from platform specific values into the proxy metadata that istiod expects today?

Reply:
At the time of this PR, we have `Metadata()` that exposed the platform metadata. However, in justin's other PR, we are adding `Labels()`, which can (carefully) be used to add to the labels that istiod expects.

If we want to further expand the Environment interface for other helpers, we definitely can. Ideally, we'd come up with an agreed upon mapping into the right places. Should we file an issue to track? Is this something you can provide some guidance on?

For this PR, let's just get _something_ in for Azure to start.